src/game.cpp

```diff
@@ -885,40 +885,71 @@ class GameOnDemandSoundFetcher: public OnDemandSoundFetcher
 	}
 };
 
+
+// before 1.8 there isn't a "integer interface", only float
+#if (IRRLICHT_VERSION_MAJOR == 1 && IRRLICHT_VERSION_MINOR < 8)
+typedef irr::f32 SamplerLayer_t;
+#else
+typedef irr::s32 SamplerLayer_t;
+#endif
+
+
 class GameGlobalShaderConstantSetter : public IShaderConstantSetter
 {
-	Sky *m_sky;
+	Sky **m_sky;
 	bool *m_force_fog_off;
 	f32 *m_fog_range;
+	bool m_fog_enabled;
+	CachedShaderSetting<float, true, 4> m_sky_bg_color;
```

Comment:
its probably okay to leave `CachedShaderSetting` as a template, but I'd prefer if not the template was used, but rather two templates derived from `CachedShaderSetting`, namely `CachedVertexShaderSetting` and `CachedPixelShaderSetting`. Then you would instantly recognize what this bool param is about.